File: BlogHybrid.Application/Services/SlugService.cs
// สร้างไฟล์ใหม่: BlogHybrid.Application/Services/SlugService.cs

using System.Text.RegularExpressions;

namespace BlogHybrid.Application.Services
{
    public interface ISlugService
    {
        string GenerateSlug(string input, int maxLength = 50);
        Task<string> GenerateUniqueSlugAsync(string input, Func<string, int?, Task<bool>> checkExists, int? excludeId = null, int maxLength = 50);
    }

    public class SlugService : ISlugService
    {
        private static readonly Dictionary<string, string> ThaiToEnglishMap = new()
        {
            // คำศัพท์ที่ใช้บ่อย (ตรวจสอบก่อนแปลงตัวอักษร)
            {"เทคโนโลยี", "technology"},
            {"ข่าวสาร", "news"},
            {"บทความ", "articles"},
            {"การศึกษา", "education"},
            {"ธุรกิจ", "business"},
            {"การเงิน", "finance"},
            {"กีฬา", "sports"},
            {"ท่องเที่ยว", "travel"},
            {"สุขภาพ", "health"},
            {"อาหาร", "food"},
            {"รถยนต์", "automotive"},
            {"บันเทิง", "entertainment"},
            {"ภาพยนตร์", "movies"},
            {"ดนตรี", "music"},
            {"แฟชั่น", "fashion"},
            {"บ้านและสวน", "home-garden"},
            {"การเมือง", "politics"},
            {"สิ่งแวดล้อม", "environment"},
            {"วิทยาศาสตร์", "science"},
            {"ประวัติศาสตร์", "history"},
            {"ศิลปะ", "art"},
            {"วรรณกรรม", "literature"},
            {"เกมส์", "games"},
            {"มือถือ", "mobile"},
            {"คอมพิวเตอร์", "computer"},
            {"อินเทอร์เน็ต", "internet"},

            // พยัญชนะไทย
            {"ก", "g"}, {"ข", "k"}, {"ค", "k"}, {"ง", "ng"},
            {"จ", "j"}, {"ฉ", "ch"}, {"ช", "ch"}, {"ซ", "s"},
            {"ญ", "y"}, {"ด", "d"}, {"ต", "t"}, {"ถ", "th"},
            {"ท", "th"}, {"ธ", "th"}, {"น", "n"}, {"บ", "b"},
            {"ป", "p"}, {"ผ", "ph"}, {"ฝ", "f"}, {"พ", "ph"},
            {"ฟ", "f"}, {"ภ", "ph"}, {"ม", "m"}, {"ย", "y"},
            {"ร", "r"}, {"ล", "l"}, {"ว", "w"}, {"ศ", "s"},
            {"ษ", "s"}, {"ส", "s"}, {"ห", "h"}, {"ฬ", "l"},
            {"อ", "o"}, {"ฮ", "h"},

            // สระไทย
            {"ะ", "a"}, {"า", "a"}, {"ิ", "i"}, {"ี", "i"},
            {"ึ", "ue"}, {"ื", "ue"}, {"ุ", "u"}, {"ู", "u"},
            {"เ", "e"}, {"แ", "ae"}, {"โ", "o"}, {"ใ", "ai"},
            {"ไ", "ai"}, {"ำ", "am"}, {"ั", "a"}, {"ํ", "am"},
            
            // เลขไทย
            {"๐", "0"}, {"๑", "1"}, {"๒", "2"}, {"๓", "3"}, {"๔", "4"},
            {"๕", "5"}, {"๖", "6"}, {"๗", "7"}, {"๘", "8"}, {"๙", "9"}
        };

        public string GenerateSlug(string input, int maxLength = 50)
        {
            if (string.IsNullOrWhiteSpace(input))
                return string.Empty;

            var slug = input.Trim().ToLowerInvariant();

            // แปลงคำศัพท์ที่ใช้บ่อยก่อน (คำยาว)
            foreach (var pair in ThaiToEnglishMap.Where(x => x.Key.Length > 1).OrderByDescending(x => x.Key.Length))
            {
                slug = slug.Replace(pair.Key, pair.Value);
            }

            // แปลงตัวอักษรทีละตัว
            foreach (var pair in ThaiToEnglishMap.Where(x => x.Key.Length == 1))
            {
                slug = slug.Replace(pair.Key, pair.Value);
            }

            // ลบวรรณยุกต์และการันต์
            slug = Regex.Replace(slug, @"[่้๊๋์]", "");

            // แทนที่ช่องว่างและอักขระพิเศษด้วย -
            slug = Regex.Replace(slug, @"[^\w\-]", "-");

            // ลบ - ที่ติดกันและที่ต้นท้าย
            slug = Regex.Replace(slug, @"-+", "-");
            slug = slug.Trim('-');

            // จำกัดความยาว
            if (slug.Length > maxLength)
            {
                slug = slug.Substring(0, maxLength).TrimEnd('-');
            }

            // ถ้าเป็นช่องว่างหรือว่างเปล่า ให้ส่งกลับ random string
            if (string.IsNullOrWhiteSpace(slug))
            {
                slug = Guid.NewGuid().ToString("N")[..8];
            }

            return slug;
        }

        public async Task<string> GenerateUniqueSlugAsync(string input, Func<string, int?, Task<bool>> checkExists, int? excludeId = null, int maxLength = 50)
        {
            var baseSlug = GenerateSlug(input, maxLength - 10); // เผื่อพื้นที่สำหรับตัวเลข
            var slug = baseSlug;
            var counter = 1;

            while (await checkExists(slug, excludeId))
            {
                slug = $"{baseSlug}-{counter}";
                counter++;

                // ป้องกันการวนซ้ำไม่สิ้นสุด
                if (counter > 999)
                {
                    slug = $"{baseSlug}-{Guid.NewGuid().ToString("N")[..8]}";
                    break;
                }
            }

            return slug;
        }
    }
}

File: BlogHybrid.Web/wwwroot/js/slugService.js
// ไฟล์: BlogHybrid.Web/wwwroot/js/slugService.js

class SlugService {
    static thaiToEnglishMap = {
        // คำศัพท์ที่ใช้บ่อย
        'เทคโนโลยี': 'technology',
        'ข่าวสาร': 'news',
        'บทความ': 'articles',
        'การศึกษา': 'education',
        'ธุรกิจ': 'business',
        'การเงิน': 'finance',
        'กีฬา': 'sports',
        'ท่องเที่ยว': 'travel',
        'สุขภาพ': 'health',
        'อาหาร': 'food',
        'รถยนต์': 'automotive',
        'บันเทิง': 'entertainment',
        'ภาพยนตร์': 'movies',
        'ดนตรี': 'music',
        'แฟชั่น': 'fashion',
        'บ้านและสวน': 'home-garden',

        // พยัญชนะไทย
        'ก': 'g', 'ข': 'k', 'ค': 'k', 'ง': 'ng',
        'จ': 'j', 'ฉ': 'ch', 'ช': 'ch', 'ซ': 's',
        'ญ': 'y', 'ด': 'd', 'ต': 't', 'ถ': 'th',
        'ท': 'th', 'ธ': 'th', 'น': 'n', 'บ': 'b',
        'ป': 'p', 'ผ': 'ph', 'ฝ': 'f', 'พ': 'ph',
        'ฟ': 'f', 'ภ': 'ph', 'ม': 'm', 'ย': 'y',
        'ร': 'r', 'ล': 'l', 'ว': 'w', 'ศ': 's',
        'ษ': 's', 'ส': 's', 'ห': 'h', 'ฬ': 'l',
        'อ': 'o', 'ฮ': 'h',

        // สระไทย
        'ะ': 'a', 'า': 'a', 'ิ': 'i', 'ี': 'i',
        'ึ': 'ue', 'ื': 'ue', 'ุ': 'u', 'ู': 'u',
        'เ': 'e', 'แ': 'ae', 'โ': 'o', 'ใ': 'ai',
        'ไ': 'ai', 'ำ': 'am', 'ั': 'a', 'ํ': 'am'
    };

    static generateSlug(input, maxLength = 50) {
        if (!input) return '';

        let slug = input.trim().toLowerCase();

        // แปลงคำศัพท์ยาวก่อน
        const longWords = Object.entries(this.thaiToEnglishMap)
            .filter(([key]) => key.length > 1)
            .sort(([a], [b]) => b.length - a.length);

        for (const [thai, eng] of longWords) {
            slug = slug.replace(new RegExp(thai, 'g'), eng);
        }

        // แปลงตัวอักษรทีละตัว
        for (const [thai, eng] of Object.entries(this.thaiToEnglishMap)) {
            if (thai.length === 1) {
                slug = slug.replace(new RegExp(thai, 'g'), eng);
            }
        }

        // ลบวรรณยุกต์
        slug = slug.replace(/[่้๊๋์]/g, '');

        // แทนที่อักขระพิเศษด้วย -
        slug = slug.replace(/[^\w\-]/g, '-');

        // ลบ - ซ้ำ
        slug = slug.replace(/-+/g, '-');
        slug = slug.replace(/^-|-$/g, '');

        // จำกัดความยาว
        if (slug.length > maxLength) {
            slug = slug.substring(0, maxLength).replace(/-$/, '');
        }

        return slug || Math.random().toString(36).substring(2, 10);
    }

    static async checkSlugExists(slug, checkUrl) {
        try {
            const response = await fetch(`${checkUrl}?slug=${encodeURIComponent(slug)}`);
            const result = await response.json();
            return result.exists || false;
        } catch (error) {
            console.error('Error checking slug:', error);
            return true;
        }
    }

    static async updateSlugStatus(slug, checkUrl, statusElement) {
        if (!slug || !statusElement) return;

        try {
            const exists = await this.checkSlugExists(slug, checkUrl);

            if (exists) {
                statusElement.innerHTML = '<span class="text-danger"><i class="fas fa-times me-1"></i>URL นี้ถูกใช้งานแล้ว</span>';
            } else {
                statusElement.innerHTML = '<span class="text-success"><i class="fas fa-check me-1"></i>URL พร้อมใช้งาน</span>';
            }
        } catch (error) {
            statusElement.innerHTML = '<span class="text-warning"><i class="fas fa-exclamation-triangle me-1"></i>ไม่สามารถตรวจสอบ URL ได้</span>';
        }
    }
}

window.SlugService = SlugService;